tests/unit/test_gce_clients.py

```diff
@@ -0,0 +1,344 @@
+                payload=compute_rsp)
+
+            results = await gcp_client.list_instances(
+                'test-project',
+                page_size=10,
+                instance_filter='zone eq us-west9-z')
+
+            expected_results = [{
+                'hostname': 'instance-1',
+                'internal_ip': '192.168.0.1',
+                'external_ip': '1.1.1.1'
+            }]
+        assert results == expected_results
+
+    @pytest.mark.asyncio
+    async def test_list_instances_filter_by_tag(
+            self, compute_rsp, compute_base_url, get_gce_client, caplog):
+        gcp_client = get_gce_client(gce_clients.GCPClient)
+        caplog.set_level(logging.ERROR)
+        blacklisted_instance = \
+            copy.deepcopy(compute_rsp['items']['us-west1-z']['instances'][0])
+        blacklisted_instance['name'] = 'instance-2'
+        blacklisted_instance['tags']['items'] = ['bl-tag']
+        compute_rsp['items']['us-west1-z']['instances'].append(
+            blacklisted_instance)
+
+        gcp_client.blacklisted_tags.append('bl-tag')
+
+        with aioresponses() as m:
+            m.get(
+                compute_base_url + 'maxResults=10&filter=',
+                payload=compute_rsp)
+
+            results = await gcp_client.list_instances(
+                'test-project', page_size=10)
+
+            expected_results = [{
+                'hostname': 'instance-1',
+                'internal_ip': '192.168.0.1',
+                'external_ip': '1.1.1.1'
+            }]
+        assert results == expected_results
+        assert len(caplog.records) == 1
+
+    @pytest.mark.asyncio
+    async def test_list_instances_filter_by_metadata(
+            self, compute_rsp, compute_base_url, get_gce_client, caplog):
+        gcp_client = get_gce_client(gce_clients.GCPClient, )
+        caplog.set_level(logging.ERROR)
+        blacklisted_instance = \
+            copy.deepcopy(compute_rsp['items']['us-west1-z']['instances'][0])
+
+        blacklisted_instance['name'] = 'instance-2'
+        blacklisted_instance['metadata']['items'].append(
+            {'key': 'type', 'value': 'c0ffee-b33f'})
+        compute_rsp['items']['us-west1-z']['instances'].append(
+            blacklisted_instance)
+
+        gcp_client.blacklisted_metadata.append({'type': 'c0ffee-b33f'})
+        with aioresponses() as m:
+            m.get(
+                compute_base_url + 'maxResults=10&filter=',
+                payload=compute_rsp)
+
+            results = await gcp_client.list_instances(
+                'test-project', page_size=10)
+
+            expected_results = [{
+                'hostname': 'instance-1',
+                'internal_ip': '192.168.0.1',
+                'external_ip': '1.1.1.1'
+            }]
+        assert results == expected_results
+        assert len(caplog.records) == 1
+
+    @pytest.mark.asyncio
+    async def test_list_instances_malformed_instance_json(
```

Comment:
This might also be able to be included in the parametrization suggested above.

Reply:
I got this working, but the resulting test was harder to understand because this case is different from the previous three and required a lot of setup specific to just this one case.